pkg/systemd/hwinfo.jsx

```diff
@@ -24,61 +24,215 @@ import React from "react";
+                        <tr>
+                            <th>{ _("CPU Security") }</th>
+                            <td>{ mitigations }</td>
+                        </tr>
+                        }
+                    </React.Fragment>
+                    }
+                </tbody>
+            </table>
+        );
+    }
+}
+
+let cachedMitigations;
+let AvailableMitigations = () => {
+    if (cachedMitigations !== undefined)
+        return cachedMitigations;
+    /* nosmt */
+    let promises = [cockpit.spawn(["lscpu"], { environ: ["LC_ALL=C.UTF-8"], }), cockpit.file("/proc/cmdline").read()];
+    return Promise.all(promises).then(values => {
+        let threads_per_core;
+        try {
+            threads_per_core = Number(values[0].split('\n')
+                    .find(l => l.indexOf('Thread(s) per core:') !== -1)
+                    .split(':')[1]);
+        } catch (e) {
+            console.warn(e);
+            return { available: false };
+        }
+        /* "nosmt" and "nosmt=force" are valid */
+        let nosmt_enabled = (values[1].indexOf("nosmt") !== -1 && values[1].indexOf("nosmt=") === -1) || values[1].indexOf("nosmt=force") !== -1;
+        /* available if threads>1 and the cmdline is valid */
+        let nosmt_available = threads_per_core > 1 && (values[1].indexOf("nosmt=") === -1 || values[1].indexOf("nosmt=force") !== -1);
+
+        cachedMitigations = { available: nosmt_available, nosmt_enabled: nosmt_enabled };
+        return cachedMitigations;
+    });
+};
+
+class CPUSecurityMitigationsDialog extends React.Component {
+    constructor(props) {
+        super(props);
+        this.close = this.close.bind(this);
+        this.saveAndReboot = this.saveAndReboot.bind(this);
+        this.state = {
+            nosmt: undefined,
+            alert: undefined,
+            mitigationsAvailable: false,
+        };
+        AvailableMitigations().then(({ available, nosmt_enabled }) => {
+            this.setState({ mitigationsAvailable: available, nosmt: nosmt_enabled });
+        });
+    }
+
+    close() {
+        if (this.props.onClose)
+            this.props.onClose();
+    }
+
+    saveAndReboot() {
+        let options = [];
+        if (this.state.nosmt)
+            options = ['set', 'nosmt'];
+        else
+            options = ['remove', 'nosmt'];
+        cockpit.script(kernelopt_sh, options, { superuser: "require" })
+                .then(() => {
+                    cockpit.spawn(["shutdown", "--reboot", "now"], { superuser: "require" })
+                            .catch(error => this.setState({ rebooting: false, alert: error }));
+                })
+                .catch(error => this.setState({ rebooting: false, alert: error }));
+    }
+
+    render() {
+        let rows = [];
+        if (this.state.nosmt !== undefined)
+            rows.push((
+                <ListView.Item key="nosmt" heading={ <span>{ _("Disable symmetric multithreading") } (nosmt)</span> }
```

Comment:
This looks fine in a narrow browser:

![dialog-narrow](https://user-images.githubusercontent.com/200109/54682902-7cfc9a00-4b10-11e9-9a66-8c545c76b833.png)

but the alignment is off in a wide one:

![dialog-wide](https://user-images.githubusercontent.com/200109/54682913-85ed6b80-4b10-11e9-81ca-6db251f24927.png)

Reply:
Im not sure if this ca nbe changed, as it's a patternfly element. We could remove text from inside the expanded item and put it in the 'small', but that's really small:

![image](https://user-images.githubusercontent.com/11140201/54697636-30c05280-4b2e-11e9-98a2-4ce97bdf4db5.png)

Reply:
The original design doesn't actually show the entire URL in the UI, just a "[icon] Read more.." link. That might fit better?

Reply:
It's not really a good idea to ever expose URLs in web apps, unless it's needed. Part of the reason is accessibility (as screen readers will read out that long URL) — but it's also considered to be bad form in general.

We could take a hybrid approach and say something like "[Read more at Red Hat's security site](#)" (this is a dummy example link) so the site is referenced in a descriptive way and the "Read" action is linked too.

Reply:
Thanks for the update! This looks a lot better now, and also gets rid of the expander.